### PR TITLE
Attn added kv processor torch 2.0 block

### PR DESCRIPTION
**Pull request by williamberman**
rebased on top of https://github.com/huggingface/diffusers/pull/3021 and https://github.com/huggingface/diffusers/pull/3011

https://github.com/huggingface/diffusers/pull/3023/commits/200a8c73a2d220ae28ecb979e0ac3e511b4a4da3 and on is the main commit

### CODE REVIEW

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -287,10 +321,38 @@ def prepare_attention_mask(self, attention_mask, target_length, batch_size=None)
...
 
-        if attention_mask.shape[0] < batch_size * head_size:
-            attention_mask = attention_mask.repeat_interleave(head_size, dim=0)
+        if out_dim == 3:
+            if attention_mask.shape[0] < batch_size * head_size:
+                attention_mask = attention_mask.repeat_interleave(head_size, dim=0)
+        elif out_dim == 4:
+            attention_mask = attention_mask.unsqueeze(1)
+            attention_mask = attention_mask.repeat_interleave(head_size, dim=1)
+
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:330]: @patrickvonplaten when using the torch built in attention and putting heads as the second dim, we need to make the attention mask also put the heads in the second dim. I'm not sure what the equivalent check for `attention_mask.shape[0] < batch_size * head_size` is. If we assume the input attention mask is always just the same batch size of the inputs, we don't have to do the check and I think this works. My understanding is that's what the original code was doing anyway since it just repeats by the head size regardless

**Review by patrickvonplaten — APPROVED** [review]

**Comment by williamberman** [comment]
waiting to merge until https://github.com/huggingface/diffusers/pull/3021 and https://github.com/huggingface/diffusers/pull/3011 are merged

File: src/diffusers/models/attention_processor.py [hunk 2]
```diff
@@ -79,22 +80,54 @@ def __init__(
...
+            else:
+                norm_cross_num_channels = cross_attention_dim
+
+            self.norm_cross = nn.GroupNorm(
+                num_channels=norm_cross_num_channels, num_groups=cross_attention_norm_num_groups, eps=1e-5, affine=True
+            )
+        else:
+            raise ValueError(
+                f"unknown cross_attention_norm: {cross_attention_norm}. Should be None, 'layer_norm' or 'group_norm'"
+            )
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:116]: Nice!

File: src/diffusers/models/attention_processor.py [hunk 3]
```diff
@@ -436,6 +509,64 @@ def __call__(self, attn: Attention, hidden_states, encoder_hidden_states=None, a
...
+            key = attn.head_to_batch_dim(key, out_dim=4)
+            value = attn.head_to_batch_dim(value, out_dim=4)
+            key = torch.cat([encoder_hidden_states_key_proj, key], dim=2)
+            value = torch.cat([encoder_hidden_states_value_proj, value], dim=2)
+        else:
+            key = encoder_hidden_states_key_proj
+            value = encoder_hidden_states_value_proj
+
+        # the output of sdp = (batch, num_heads, seq_len, head_dim)
+        # TODO: add support for attn.scale when we move to Torch 2.1
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:553]: Out of curiosity. Which scale is this one?

File: src/diffusers/models/attention_processor.py [hunk 4]
```diff
@@ -287,10 +324,35 @@ def prepare_attention_mask(self, attention_mask, target_length, batch_size=None)
...
+            encoder_hidden_states = self.norm_cross(encoder_hidden_states)
+        elif isinstance(self.norm_cross, nn.GroupNorm):
+            # Group norm norms along the channels dimension and expects
+            # input to be in the shape of (N, C, *). In this case, we want
+            # to norm along the hidden dimension, so we need to move
+            # (batch_size, sequence_length, hidden_size) ->
+            # (batch_size, hidden_size, sequence_length)
+            encoder_hidden_states = encoder_hidden_states.transpose(1, 2)
+            encoder_hidden_states = self.norm_cross(encoder_hidden_states)
+            encoder_hidden_states = encoder_hidden_states.transpose(1, 2)
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:350]: 👌

File: src/diffusers/models/attention_processor.py [hunk 5]
```diff
@@ -403,23 +468,31 @@ def __call__(self, attn: Attention, hidden_states, encoder_hidden_states=None, a
...
+        if not attn.only_cross_attention:
+            key = attn.to_k(hidden_states)
+            value = attn.to_v(hidden_states)
+            key = attn.head_to_batch_dim(key)
+            value = attn.head_to_batch_dim(value)
+            key = torch.cat([encoder_hidden_states_key_proj, key], dim=1)
+            value = torch.cat([encoder_hidden_states_value_proj, value], dim=1)
+        else:
+            key = encoder_hidden_states_key_proj
+            value = encoder_hidden_states_value_proj
```
Comment (sayakpaul) [inline src/diffusers/models/attention_processor.py:495]: Clean 🌿

File: src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py [hunk 6]
```diff
@@ -1502,7 +1520,9 @@ def __init__(
                     norm_num_groups=resnet_groups,
                     bias=True,
                     upcast_softmax=True,
-                    processor=AttnAddedKVProcessor(),
+                    only_cross_attention=only_cross_attention,
+                    cross_attention_norm=cross_attention_norm,
+                    processor=processor,
```
Comment (sayakpaul) [inline src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py:1525]: 👌

File: tests/pipelines/unclip/test_unclip_image_variation.py [hunk 7]
```diff
@@ -421,7 +421,12 @@ class DummyScheduler:
     def test_attention_slicing_forward_pass(self):
         test_max_difference = torch_device == "cpu"
 
-        self._test_attention_slicing_forward_pass(test_max_difference=test_max_difference)
+        # Check is relaxed because there is not a torch 2.0 sliced attention added kv processor
+        expected_max_diff = 1e-2
+
+        self._test_attention_slicing_forward_pass(
+            test_max_difference=test_max_difference, expected_max_diff=expected_max_diff
+        )
```
Comment (sayakpaul) [inline tests/pipelines/unclip/test_unclip_image_variation.py:429]: Should this be conditioned on the torch version being used?

**Review by sayakpaul — COMMENTED** [review]
Very nice refactor!

File: tests/pipelines/unclip/test_unclip_image_variation.py (hunk 7 above)
Comment (williamberman) [reply tests/pipelines/unclip/test_unclip_image_variation.py:429]: Sorry, intention is to follow up and fix the test data
